docs/developer-docs/security/security-best-practices/data-integrity-and-authenticity.mdx

```diff
@@ -8,19 +8,324 @@ import { MarkdownChipRow } from "/src/components/Chip/MarkdownChipRow";
 
-The responses to [query calls](/docs/current/references/ic-interface-spec#http-interface) are not threshold-signed by the canister and subnet. Thus, a single malicious replica or boundary node may change the data, violating its authenticity. This is especially risky if update calls, which do returned signed responses, depend on the response to query calls.
+:::tip
+ICP also provides Replica signed queries, where query responses are signed by the answering replica node but however it doesn't have the same security guarantees as `update` calls. Replica signed queries are enabled by default on both `agent-rs` and `agent-js`. You can read more about the feature [here](https://internetcomputer.org/blog/features/replica-signed-queries)
+:::
 
-### Recommendation
+### What's Certified data?
+* A canister can store a small amount of data (32 bytes) during an update call in the state tree under `certified_data` which will be certified by the subnet.
+* During query call, the canister can obtain a certificate about that data, which can be verified by the IC root public key.
+
+In the rest of the section, we will see an example of how to design a canister, which can serve a certified response for a `query` using `certified_data` and how to verify the response. The examples are written in Rust but can be imported to other languages following the same recommendations.
+
+### Building a canister with Certified Variables
+Let's consider the following canister,
+
+```c
+type User = record {
+    name: text;
+    age: nat8;
+};
+
+type CertifiedUser = record {
+    user : User;
+    certificate : blob;
+    witness : blob;
+};
+
+service : {
+    "set_user": (User) -> (nat64);
+    "get_user": (nat64) -> (CertifiedUser) query;
+}
+```
+
+The canister exposes the following service, 
+* **set_user** - The caller provides a `User` object to the canister. The canister records it and serves a corresponding `index` for the entry as the response. Since `certified_data` can only store 32 bytes of data, we use a specialized data strucutre from `ic_certified_map` to store the `User` data. 
```

Comment:
```suggestion
* **set_user**: The caller provides a `User` object to the canister. The canister records it and serves a corresponding `index` for the entry as the response. Since `certified_data` can only store 32 bytes of data, it uses a specialized data structure from `ic_certified_map` to store the `User` data. 
```